### PR TITLE
[RLlib] Issue 25696: Output writers not working w/ multiple workers.

### LINKED ISSUE #25696: [RLlib]: Output writers not working with multiple workers
### What happened + What you expected to happen

I'm running into a bug using multiple rollout_workers and outputting the data to json either directly or using the DatasetWriter.

Traceback:
```
== Status ==
Current time: 2022-06-10 23:32:20 (running for 00:00:09.60)
Memory usage on this node: 10.7/22.9 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/1 GPUs, 0.0/8.57 GiB heap, 0.0/4.28 GiB objects (0.0/1.0 accelerator_type:G)
Result logdir: /home/rohan/ray_results/default
Number of trials: 1/1 (1 ERROR)
Number of errored trials: 1
+----------------------------+--------------+--------------------------------------------------------------------------------------------+
| Trial name                 |   # failures | error file                                                                                 |
|----------------------------+--------------+--------------------------------------------------------------------------------------------|
| PG_CartPole-v0_390ed_00000 |            1 | /home/rohan/ray_results/default/PG_CartPole-v0_390ed_00000_0_2022-06-10_23-32-10/error.txt |
+----------------------------+--------------+--------------------------------------------------------------------------------------------+

== Status ==
Current time: 2022-06-10 23:32:20 (running for 00:00:09.60)
Memory usage on this node: 10.6/22.9 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/1 GPUs, 0.0/8.57 GiB heap, 0.0/4.28 GiB objects (0.0/1.0 accelerator_type:G)
Result logdir: /home/rohan/ray_results/default
Number of trials: 1/1 (1 ERROR)
Number of errored trials: 1
+----------------------------+--------------+--------------------------------------------------------------------------------------------+
| Trial name                 |   # failures | error file                                                                                 |
|----------------------------+--------------+--------------------------------------------------------------------------------------------|
| PG_CartPole-v0_390ed_00000 |            1 | /home/rohan/ray_results/default/PG_CartPole-v0_390ed_00000_0_2022-06-10_23-32-10/error.txt |
+----------------------------+--------------+--------------------------------------------------------------------------------------------+

2022-06-10 23:32:20,414 ERROR ray_trial_executor.py:107 -- An exception occurred when trying to stop the Ray actor:Traceback (most recent call last):
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/ray_trial_executor.py", line 98, in post_stop_cleanup
    ray.get(future, timeout=0)
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/worker.py", line 2073, in get
    raise value
ray.exceptions.RayActorError: The actor died because of an error raised in its creation task, ray::PG.__init__() (pid=345635, ip=10.0.0.105, repr=PG)
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 293, in __init__
    super().__init__(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/trainable.py", line 156, in __init__
    self.setup(copy.deepcopy(self.config))
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 388, in setup
    self.workers = WorkerSet(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/rllib/evaluation/worker_set.py", line 127, in __init__
    self.add_workers(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/rllib/evaluation/worker_set.py", line 270, in add_workers
    self.foreach_worker_with_index(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/rllib/evaluation/worker_set.py", line 405, in foreach_worker_with_index
    remote_results = ray.get(
ray.exceptions.RayTaskError(TypeError): ray::RolloutWorker.apply() (pid=345766, ip=10.0.0.105, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x7ff4629b9d30>)
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 421, in serialize
    return self._serialize_to_msgpack(value)
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 399, in _serialize_to_msgpack
    pickle5_serialized_object = self._serialize_to_pickle5(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 361, in _serialize_to_pickle5
    raise e
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 356, in _serialize_to_pickle5
    inband = pickle.dumps(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 620, in dump
    return Pickler.dump(self, obj)
TypeError: cannot pickle 'generator' object

(PG pid=345635) 2022-06-10 23:32:20,405 ERROR worker.py:289 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::RolloutWorker.apply() (pid=345768, ip=10.0.0.105, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x7f39e1c23d30>)
(PG pid=345635)   File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 421, in serialize
(PG pid=345635)     return self._serialize_to_msgpack(value)
(PG pid=345635)   File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 399, in _serialize_to_msgpack
(PG pid=345635)     pickle5_serialized_object = self._serialize_to_pickle5(
(PG pid=345635)   File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 361, in _serialize_to_pickle5
(PG pid=345635)     raise e
(PG pid=345635)   File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/serialization.py", line 356, in _serialize_to_pickle5
(PG pid=345635)     inband = pickle.dumps(
(PG pid=345635)   File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 73, in dumps
(PG pid=345635)     cp.dump(obj)
(PG pid=345635)   File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 620, in dump
(PG pid=345635)     return Pickler.dump(self, obj)
(PG pid=345635) TypeError: cannot pickle 'generator' object
(RolloutWorker pid=345768) /home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/gym/envs/registration.py:564: UserWarning: WARN: The environment CartPole-v0 is out of date. You should consider upgrading to version `v1`.
(RolloutWorker pid=345768)   logger.warn(
(RolloutWorker pid=345768) /home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/gym/utils/passive_env_checker.py:97: UserWarning: WARN: We recommend you to use a symmetric and normalized Box action space (range=[-1, 1]) https://stable-baselines3.readthedocs.io/en/master/guide/rl_tips.html
(RolloutWorker pid=345768)   logger.warn(
Traceback (most recent call last):
  File "/home/rohan/miniconda3/envs/ray/bin/rllib", line 8, in <module>
    sys.exit(cli())
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/rllib/scripts.py", line 41, in cli
    train.run(options, train_parser)
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/rllib/train.py", line 269, in run
    run_experiments(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/tune.py", line 831, in run_experiments
    return run(
  File "/home/rohan/miniconda3/envs/ray/lib/python3.8/site-packages/ray/tune/tune.py", line 735, in run
    raise TuneError("Trials did not complete", incomplete_trials)
ray.tune.error.TuneError: ('Trials did not complete', [PG_CartPole-v0_390ed_00000])
```

### Versions / Dependencies

ray version 3.0.0.dev0 (nightly build wheel, then set up symlinks with `python setup-dev.py`)

### Reproduction script

`rllib train --run=PG --env=CartPole-v0 --framework=torch --config='{"num_workers": 8, "output": "/tmp/cp-out/"}' --stop='{"episodes_total": 10000}'`

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by sven1977**
Fixes issue:
https://github.com/ray-project/ray/issues/25696

## Why are these changes needed?
Issue #25696 

## Related issue number
Closes #25696 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/worker_set.py [hunk 1]
```diff
@@ -267,9 +267,7 @@ def add_workers(self, num_workers: int, validate: bool = False) -> None:
         # which needs to be handled by this WorkerSet's owner (usually
         # a RLlib Algorithm instance).
         if validate:
-            self.foreach_worker_with_index(
-                lambda w, i: w.policy_map and w.input_reader and w.output_writer
-            )
+            self.foreach_worker(lambda w: w.assert_healthiness())
```
Comment (gjoliver) [inline rllib/evaluation/worker_set.py:270]: I am curious, can you quickly try if it works to put assert in the lambda, not in the function directly?
then, we can make the function is_healthy(), and may potentially be more useful.

Comment (Rohan138) [reply rllib/evaluation/worker_set.py]: Yup, just changed it, ready to merge

Comment (Rohan138) [reply rllib/evaluation/worker_set.py]: Hmm, I'm okay either way, should we merge this? (Either before or after the last commit)

Comment (Rohan138) [reply rllib/evaluation/worker_set.py]: Yeah, you can't use assert in a lambda. [Link](https://stackoverflow.com/questions/8477346/checking-assertions-in-a-lambda-in-python)

Still don't think that's the underlying issue, though, but I haven't been able to come up with an MWP that fails.

I added a test case to `test_io.py` to make sure it works correctly.

**Review by gjoliver — APPROVED** [review]
thanks

**Pull request merged**